File: Ejercicio 1/ej1_nuevo.cpp
#include <iostream>
#include <cassert>
#include <stdlib.h>
#include <stdio.h>
#include <sstream>
#include <string>
#include <set>

#include "nodo.h"
#include "ej1.h"

using namespace std;


void split(const string &s, char delim, vector<int> &elems) {
    stringstream ss;
    ss.str(s);
    string item;
    while (getline(ss, item, delim)) {
        elems.push_back(atoi(item.c_str()));
    }
}

struct Estado {
 	Estado();
 	Estado(vector<unsigned int>&a, vector<unsigned int>&b, vector<unsigned int>&c, vector<unsigned int>&d, bool i) {
 		ida = i;
 		if (i) {
 			iA = set<unsigned int>(a.begin(), a.end());
 			iB = set<unsigned int>(b.begin(), b.end());
 			cA = set<unsigned int>(c.begin(), c.end());
 			cB = set<unsigned int>(d.begin(), d.end());
 		} else {
 			iA = set<unsigned int>(b.begin(), b.end());
 			iB = set<unsigned int>(a.begin(), a.end());
 			cA = set<unsigned int>(d.begin(), d.end());
 			cB = set<unsigned int>(c.begin(), c.end());
 		}
 	}

 	set<unsigned int> iA;
 	set<unsigned int> iB;
 	set<unsigned int> cA;
 	set<unsigned int> cB;
 	bool ida;

 	bool iguales(Estado& otro) {
 		return (iA == otro.iA && iB == otro.iB && cA == otro.cA && cB == otro.cB && ida == otro.ida);
 	}

 };

bool nodo_visitado(vector<Estado>& movimientos, vector<unsigned int>& iA, vector<unsigned int>& iB, vector<unsigned int>& cA, vector<unsigned int>& cB, bool ida) {
	Estado e = Estado(iA, iB, cA, cB, ida);
	for (int i = 0; i < movimientos.size(); i++) {
		if (movimientos[i].iguales(e)) return true;
	}
	return false;
}

void agregar_nodo(vector<Estado>& movimientos, vector<unsigned int>& iA, vector<unsigned int>& iB, vector<unsigned int>& cA, vector<unsigned int>& cB, bool ida) {
	movimientos.push_back(Estado(iA, iB, cA, cB, ida));
}

// Lo declaro primero porque si no no compila. La funcion ida() la usa y si no esta arriba, no sabe que es.
Nodo* cruzar_puente(vector<unsigned int> indysA, vector<unsigned int> canibA, vector<unsigned int> indysB, vector<unsigned int> canibB, bool ida, vector<Estado> movimientos, int vel, int* velocidad_minima) {
	// No importa cual sea el escenario siempre voy a necesitar un nodo nuevo
	Nodo* new_nodo = new Nodo(); // Por default su vel = 0 y valido = true
	
	// Si hay mas canibales que arqueologos de algun lado (y cantidad de arqueologos != 0) perdi
	if ((!indysA.empty() and indysA.size() < canibA.size()) || (!indysB.empty() and indysB.size() < canibB.size())) {
		new_nodo->valido = false; // Marco en el nodo que NO es solucion
		return new_nodo;

	// Si no hay mas arqueologos ni canibales del lado A, gane
	} else if ((indysA.empty() && canibA.empty() && ida) || (indysB.empty() && canibB.empty() && !ida)) {
		if (*velocidad_minima != -1) *velocidad_minima = min(*velocidad_minima, vel); // Seteo su velocidad
		else *velocidad_minima = vel;
		return new_nodo;
	} else if (*velocidad_minima != -1 && vel > *velocidad_minima) {
		new_nodo->valido = false; // Marco en el nodo que NO es solucion
		return new_nodo;
	}

	for (unsigned int i = 0; i < indysA.size(); i++) {
		// copio los vectores para modificarlos sin perder los originales
		vector<unsigned int> iA (indysA);
		vector<unsigned int> iB (indysB);

		unsigned int new_vel = vel + iA[i]; // sumo a la velocidad actual el maximo entre los dos arqueologos

		// Los agrego al vector del otro lado
		iB.push_back(iA[i]);

		// los saco del vector
		iA.erase(iA.begin() + i);

		if (!nodo_visitado(movimientos, iA, iB, canibA, canibB, !ida)) {
			agregar_nodo(movimientos, iA, iB, canibA, canibB, !ida);
			Nodo* res = cruzar_puente(iB, canibB, iA, canibA, !ida, movimientos, new_vel, velocidad_minima); // hago la recursion
			movimientos.pop_back();
			if (res->valido) // si es solucion
				new_nodo->a.push_back(res); // lo guardo en el nodo nuevo que va a devolver la funcion
		}

		for (unsigned int j = i + 1; j < indysA.size(); j++) {

			// copio los vectores para modificarlos sin perder los originales
			vector<unsigned int> iA (indysA);
			vector<unsigned int> iB (indysB);

			unsigned int new_vel = vel + max(iA[i], iA[j]); // sumo a la velocidad actual el maximo entre los dos arqueologos

			// Los agrego al vector del otro lado
			iB.push_back(iA[i]);
			iB.push_back(iA[j]);

			// los saco del vector
			iA.erase(iA.begin() + i);
			iA.erase(iA.begin() + (j - 1)); // el -1 es porque ya sace el iA[i] entonces hay un elemento menos

			if (!nodo_visitado(movimientos, iA, iB, canibA, canibB, !ida)) {
				agregar_nodo(movimientos, iA, iB, canibA, canibB, !ida);
				Nodo* res = cruzar_puente(iB, canibB, iA, canibA, !ida, movimientos, new_vel, velocidad_minima); // hago la recursion
				movimientos.pop_back();
				if (res->valido) // si es solucion
					new_nodo->aa.push_back(res); // lo guardo en el nodo nuevo que va a devolver la funcion
			}
		}
	}

	for (unsigned int i = 0; i < canibA.size(); i++) {
		vector<unsigned int> cA (canibA);
		vector<unsigned int> cB (canibB);

		unsigned int new_vel = vel + cA[i]; // sumo a la velocidad actual el maximo entre los dos arqueologos

		// Los agrego al vector del otro lado
		cB.push_back(cA[i]);

		// los saco del vector
		cA.erase(cA.begin() + i);

		if (!nodo_visitado(movimientos, indysA, indysB, cA, cB, !ida)) {
			agregar_nodo(movimientos, indysA, indysB, cA, cB, !ida);
			Nodo* res = cruzar_puente(indysB, cB, indysA, cA, !ida, movimientos, new_vel, velocidad_minima); // hago la recursion
			movimientos.pop_back();
			if (res->valido) // si es solucion
				new_nodo->c.push_back(res); // lo guardo en el nodo nuevo que va a devolver la funcion
		}
		for (unsigned int j = i + 1; j < canibA.size(); j++) {
			vector<unsigned int> cA (canibA);
			vector<unsigned int> cB (canibB);

			unsigned int new_vel = vel + max(cA[i], cA[j]);

			cB.push_back(cA[i]);
			cB.push_back(cA[j]);

			cA.erase(cA.begin() + i);
			cA.erase(cA.begin() + (j - 1));

			if (!nodo_visitado(movimientos, indysA, indysB, cA, cB, !ida)) {
				agregar_nodo(movimientos, indysA, indysB, cA, cB, !ida);
				Nodo* res = cruzar_puente(indysB, cB, indysA, cA, !ida, movimientos, new_vel, velocidad_minima); // hago la recursion
				movimientos.pop_back();
				if (res->valido) // si es solucion
					new_nodo->cc.push_back(res); // lo guardo en el nodo nuevo que va a devolver la funcion
			}
		}
	}

	for (unsigned int i = 0; i < indysA.size(); i++) {
		for (unsigned int j = 0; j < canibA.size(); j++) {
			// copio todos los vectores para poder modificarlos sin perder los originales (que voy a necesitar para cada iteracion)
			vector<unsigned int> iA (indysA);
			vector<unsigned int> iB (indysB);
			vector<unsigned int> cA (canibA);
			vector<unsigned int> cB (canibB);

			unsigned int new_vel = vel + max(iA[i], cA[j]); // nueva velocidad

			// los meto en su respectivos vectores contrarios
			iB.push_back(iA[i]);
			cB.push_back(cA[j]);

			// los elimino de sus vectores originales
			iA.erase(iA.begin() + i);
			cA.erase(cA.begin() + j);

			if (!nodo_visitado(movimientos, iA, iB, cA, cB, !ida)) {
				agregar_nodo(movimientos, iA, iB, cA, cB, !ida);
				Nodo* res = cruzar_puente(iB, cB, iA, cA, !ida, movimientos, new_vel, velocidad_minima); // hago la recursion
				movimientos.pop_back();
				if (res->valido) // si es solucion
					new_nodo->ac.push_back(res); // lo guardo en el nodo nuevo que va a devolver la funcion
			}
		}
	}
	// si todos sus vectores de ramas estan vacios quiere decir que ninguno fue solucion, entonces este nodo tampoco lo es
	// caso contrario, por default valido = true
	new_nodo->valido = !(new_nodo->aa.empty() and new_nodo->cc.empty() and new_nodo->ac.empty() and new_nodo->c.empty() and new_nodo->a.empty());
	return new_nodo;
}

int main(int argc, char const *argv[]) {
	
	vector<unsigned int> indysA;
	vector<unsigned int> indysB;

	vector<unsigned int> canibA;
	vector<unsigned int> canibB;

	unsigned int cantidad_arqueologos;
	unsigned int cantidad_canibales;

	string entrada;
	getline(cin, entrada, '\n'); // tomo toda la linea de entrada como string

	// saco el 1er y 3er elemento, esto es medio harcodeado por dos cosas
	// 1. asumo que no me van a pasar numeros mayores a 6 y por lo tanto no van a tener dos digitos, por eso [0] y [2] y no algo mas unsigned int
	// 2. como cada numero esta en 'char' tengo que pasarlo a 'unsigned int', '0' = 48. Ej. '2' = 50, 2 = '2' - 48
	cantidad_arqueologos = (unsigned int) entrada[0] -48;
	cantidad_canibales = (unsigned int) entrada[2] -48;

	assert(cantidad_arqueologos > 0 and cantidad_arqueologos >= cantidad_canibales); // un assert para no pifiarla
	std::vector<int> numeros;

	if (cantidad_arqueologos > 0) {
		cout << "Ingrese " << cantidad_arqueologos << " arqueologos: ";
		getline(cin, entrada, '\n');

		split(entrada, ' ', numeros);

		for (unsigned int i = 0; i < numeros.size(); i++) {
			indysA.push_back(numeros[i]); // aca la idea es la misma que arriba
		}
	}

	assert(cantidad_arqueologos == indysA.size());
	numeros.clear();

	if (cantidad_canibales > 0) {
		cout << "Ingrese " << cantidad_canibales << " canibales: ";
		getline(cin, entrada, '\n');
		split(entrada, ' ', numeros);

		for (unsigned int i = 0; i < numeros.size(); i++) {
			canibA.push_back(numeros[i]); // aca la idea es la misma que arriba
		}
	}

	assert(cantidad_canibales == canibA.size());

	vector<Estado> movimientos;
	int velocidad = -1;

	Nodo* res = cruzar_puente(indysA, canibA, indysB, canibB, true, movimientos, 0, &velocidad);
	cout << velocidad << endl;

	delete res; // no hay que olvidar de deletear que si no perdemos memoria

	return 0;
}
	

File: Ejercicio 1/ej1_nuevo.py
import sys

global velocidad_minima
velocidad_minima = 999999999

class Nodo():
	def __init__(self):
		self.a = []
		self.c = []
		self.aa = []
		self.cc = []
		self.ac = []

class Estado():
	def __init__(self, iA, iB, cA, cB, ida):
		self.iA = set(iA)
		self.iB = set(iB)
		self.cA = set(cA)
		self.cB = set(cB)
		self.ida = ida

	def iguales(self, otro):
		return (self.iA == otro.iA and self.iB == otro.iB and self.cA == otro.cA and self.cB == otro.cB and self.ida == otro.ida)
		
def existe_estado(movimientos, iA, iB, cA, cB, ida):
	e = Estado(iA, iB, cA, cB, ida)
	for est in movimientos:
		if est.iguales(e): return True
	return False

def agregar_estado(movimientos, iA, iB, cA, cB, ida):
	movimientos.append(Estado(iA, iB, cA, cB, ida))

def ida(indysA, indysB, canivA, canivB, movimientos, mov, vel):
	global velocidad_minima
	
	if velocidad_minima < vel:
		return (False, Nodo(), -1)		

	if (len(indysA) > 0 and (len(indysA) < len(canivA))) or (len(indysB) > 0 and (len(indysB) < len(canivB))):
		return (False, Nodo(), -1)
	elif len(indysA) == 0 and len(canivA) == 0:
		# print movimientos, vel
		velocidad_minima = min(velocidad_minima, vel)
		print velocidad_minima
		return (True, Nodo(), vel)

	velMin	= 9999999999
	nodo = Nodo()

	for i in xrange(0,len(indysA)):
		iA = list(indysA) # pasa por copia
		iB = list(indysB)

		newVel = vel + iA[i]
		iB.append(iA[i])

		mov.append('A' + str(iA[i]))
		del iA[i]

		if not existe_estado(movimientos, iA, iB, canivA, canivB, 1):
			agregar_estado(movimientos, iA, iB, canivA, canivB, 1)
			res = vuelta(iA, iB, canivA, canivB, movimientos, mov, newVel)
			movimientos.pop()
			if res[0] == 1:
				nodo.a.append(res[1])
				velMin = min(velMin, res[2])
	
		mov.pop()

		for j in xrange(i+1, len(indysA)):
			iA = list(indysA) # pasa por copia
			iB = list(indysB)

			newVel = vel + max(iA[i], iA[j])
			iB.append(iA[i])
			iB.append(iA[j])

			mov.append('A' + str(iA[i]) + 'A' + str(iA[j]))
			del iA[i]
			del iA[j-1]

			if not existe_estado(movimientos, iA, iB, canivA, canivB, 1):
				agregar_estado(movimientos, iA, iB, canivA, canivB, 1)
				res = vuelta(iA, iB, canivA, canivB, movimientos, mov, newVel)
				movimientos.pop()
				if res[0] == 1:
					nodo.aa.append(res[1])
					velMin = min(velMin, res[2])
		
			mov.pop()
			pass
		
		pass

	for i in xrange(0,len(canivA)):
		cA = list(canivA) # pasa por copia
		cB = list(canivB)

		newVel = vel + cA[i]
		cB.append(cA[i])

		mov.append('C' + str(cA[i]))
		del cA[i]

		if not existe_estado(movimientos, indysA, indysB, cA, cB, 1):
			agregar_estado(movimientos, indysA, indysB, cA, cB, 1)
			res = vuelta(indysA, indysB, cA, cB, movimientos, mov, newVel)
			movimientos.pop()
			if res[0] == 1:
				nodo.c.append(res[1])
				velMin = min(velMin, res[2])
	
		mov.pop()

		for j in xrange(i+1, len(canivA)):
			cA = list(canivA) # pasa por copia
			cB = list(canivB)

			newVel = vel + max(cA[i], cA[j])
			cB.append(cA[i])
			cB.append(cA[j])

			mov.append('C' + str(cA[i]) + 'C' + str(cA[j]))
			del cA[i]
			del cA[j-1]

			if not existe_estado(movimientos, indysA, indysB, cA, cB, 1):
				agregar_estado(movimientos, indysA, indysB, cA, cB, 1)
				res = vuelta(indysA, indysB, cA, cB, movimientos, mov, newVel)
				movimientos.pop()
				if res[0] == 1:
					nodo.cc.append(res[1])
					velMin = min(velMin, res[2])
		
			mov.pop()
			pass
		pass


	if len(indysA) > 0 and len(canivA) > 0:		
		for i in xrange(0,len(indysA)):
			for j in xrange(0,len(canivA)):

				iA = list(indysA)
				iB = list(indysB)
				cA = list(canivA)
				cB = list(canivB)

				newVel = vel + max(iA[i], cA[j])
				iB.append(iA[i])
				cB.append(cA[j])

				mov.append('A' + str(iA[i]) + ' C' + str(cA[j]))
				del iA[i]
				del cA[j]
				if not existe_estado(movimientos, iA, iB, cA, cB, 1):
					agregar_estado(movimientos, iA, iB, cA, cB, 1)
					res = vuelta(iA, iB, cA, cB, movimientos, mov, newVel)
					movimientos.pop()
					if res[0] == 1:
						nodo.ac.append(res[1])
						velMin = min(velMin, res[2])
		
				mov.pop()
			pass
		pass

	valido = nodo.aa != [] or nodo.cc != [] or nodo.a != [] or nodo.c != [] or nodo.ac != []
	return (valido, nodo, velMin)



def vuelta(indysA, indysB, canivA, canivB, movimientos, mov, vel):
	global velocidad_minima
	
	if velocidad_minima < vel:
		return (False, Nodo(), -1)		

	if (len(indysA) > 0 and (len(indysA) < len(canivA))) or (len(indysB) > 0 and (len(indysB) < len(canivB))):
		return (False, Nodo(), vel)
	elif len(indysA) == 0 and len(canivA) == 0:
		# print mov, vel
		velocidad_minima = min(velocidad_minima, vel)
		return (True, Nodo(), vel)

	
	velMin	= 9999999999
	nodo = Nodo()

	for i in xrange(0,len(indysB)):
		iB = list(indysB) # pasa por copia
		iA = list(indysA)

		newVel = vel + iB[i]
		iA.append(iB[i])

		mov.append('A' + str(iB[i]))
		del iB[i]

		if not existe_estado(movimientos, iA, iB, canivA, canivB, 0):
			agregar_estado(movimientos, iA, iB, canivA, canivB, 0)
			res = ida(iA, iB, canivA, canivB, movimientos, mov, newVel)
			movimientos.pop()
			if res[0] == 1:
				nodo.a.append(res[1])
				velMin = min(velMin, res[2])
	
		mov.pop()

		for j in xrange(i+1, len(indysB)):
			iB = list(indysB) # pasa por copia
			iA = list(indysA)

			newVel = vel + max(iB[i], iB[j])
			iA.append(iB[i])
			iA.append(iB[j])

			mov.append('A' + str(iB[i]) + 'A' + str(iB[j]))
			del iB[i]
			del iB[j-1]

			if not existe_estado(movimientos, iA, iB, canivA, canivB, 0):
				agregar_estado(movimientos, iA, iB, canivA, canivB, 0)
				res = ida(iA, iB, canivA, canivB, movimientos, mov, newVel)
				movimientos.pop()
				if res[0] == 1:
					nodo.aa.append(res[1])
					velMin = min(velMin, res[2])
		
			mov.pop()
			pass
		
		pass

	for i in xrange(0,len(canivB)):
		cA = list(canivA) # pasa por copia
		cB = list(canivB)

		newVel = vel + cB[i]
		cA.append(cB[i])

		mov.append('C' + str(cB[i]))
		del cB[i]

		if not existe_estado(movimientos, indysA, indysB, cA, cB, 0):
			agregar_estado(movimientos, indysA, indysB, cA, cB, 0)
			res = ida(indysA, indysB, cA, cB, movimientos, mov, newVel)
			movimientos.pop()
			if res[0] == 1:
				nodo.c.append(res[1])
				velMin = min(velMin, res[2])
	
		mov.pop()

		for j in xrange(i+1, len(canivB)):
			cA = list(canivA) # pasa por copia
			cB = list(canivB)

			newVel = vel + max(cB[i], cB[j])
			cA.append(cB[i])
			cA.append(cB[j])

			mov.append('C' + str(cB[i]) + 'C' + str(cB[j]))
			del cB[i]
			del cB[j-1]

			if not existe_estado(movimientos, indysA, indysB, cA, cB, 0):
				agregar_estado(movimientos, indysA, indysB, cA, cB, 0)
				res = ida(indysA, indysB, cA, cB, movimientos, mov, newVel)
				movimientos.pop()
				if res[0] == 1:
					nodo.cc.append(res[1])
					velMin = min(velMin, res[2])
		
			mov.pop()
			pass
		pass


	if len(indysB) > 0 and len(canivB) > 0:		
		for i in xrange(0,len(indysB)):
			for j in xrange(0,len(canivB)):

				iA = list(indysA)
				iB = list(indysB)
				cA = list(canivA)
				cB = list(canivB)

				newVel = vel + max(iB[i], cB[j])
				iA.append(iB[i])
				cA.append(cB[j])

				mov.append('A' + str(iB[i]) + ' C' + str(cB[j]))
				del iB[i]
				del cB[j]
				if not existe_estado(movimientos, iA, iB, cA, cB, 0):
					agregar_estado(movimientos, iA, iB, cA, cB, 0)
					res = ida(iA, iB, cA, cB, movimientos, mov, newVel)
					movimientos.pop()
					if res[0] == 1:
						nodo.ac.append(res[1])
						velMin = min(velMin, res[2])
		
				mov.pop()
			pass
		pass

	valido = nodo.aa != [] or nodo.cc != [] or nodo.a != [] or nodo.c != [] or nodo.ac != []
	return (valido, nodo, velMin)


indysA = [1, 2, 3]
canivA = [4, 5, 6]

res = ida(indysA, [], canivA, [], [], [], 0)

print res[2], velocidad_minima